src/adr/0023-cassandra-mmailbox-counters-inconsistencies.md

```diff
@@ -0,0 +1,62 @@
+# 22. Cassandra Message inconsistencies
+
+Date: 2020-03-07
+
+## Status
+
+Proposed
+
+## Context
+
+Cassandra maintains a per mailbox projection for message count and unseen message count.
+
+As with any projection, it can go out of sync, leading to inconsistent results being returned to the client, which is not acceptable.
+
+Here is the table organisation:
+
+ - `mailbox` Lists the mailboxes
+ - `messageIdTable` Holds mailbox and flags for each message, lookup by mailbox ID + UID
+ - `imapUidTable` Holds mailbox and flags for each message, lookup by message ID and serves as a source of truth
+ - `mailboxCounters` Holds messages count and unseen message count for each mailbox.
+ 
+Failures during the denormalization process will lead to inconsistencies between the counts and the content of `imapUidTable`
+
+This can lead to the following user experience:
+
+ - Invalid message count can be reported in the Mail User Agent (IMAP & JMAP)
+ - Invalid message unseen count can be reported in the Mail User Agent (IMAP & JMAP)
+
+## Decision
+
+Implement a webadmin exposed task to recompute mailbox counters.
+
+This endpoints will:
+
+ - List existing mailboxes
+ - List their messages using `messageIdTable`
+ - Check them against their source of truth `imapUidTable`
+ - Compute mailbox counter values
+ - And reset the value of the counter if needed in `mailboxCounters`
+
+## Consequences
+
+This endpoint won't be friendly in the face of concurrent operations, which would be ignored during the recomputation of
```

Comment:
Phrasing

Reply:
?

Reply:
In case of concurrent operations, the endpoint's request will be ignored during the recomputation of...

Reply:
This endpoint won't be friendly in the face of concurrent operations. Concurrent increme,t & decrements will be ignored during a single mailbox processing.

Would that be Ok for you?

Reply:
*increment*

replace the friendly and we're good